hypervisor/virt/hyperv/hypervcli.py

```diff
@@ -101,4 +101,139 @@ def host_uuid(self):
         output = self._format(ret, output)[0]
         return output["UUID"]
 
+    def host_switch(self):
+        """
+        Get the switch name for hyperv manager
+        :return: the switch name
+        """
+        cmd = "PowerShell ConvertTo-Json @(Get-VMSwitch)"
+        ret, output = self.ssh.runcmd(cmd)
+        output = self._format(ret, output)[0]
+        return output['Name']
+
+    def guest_image(self, guest_name, image_path):
+        cmd = 'PowerShell New-Item -path C:\ -Name hyperv_img -Type Directory'
+        self.ssh.runcmd(cmd)
+        cmd = 'PowerShell Get-ChildItem C:\hyperv_img'
+        ret, output = self.ssh.runcmd(cmd)
+        if not ret and guest_name in output:
+            logger.info("hyperv image is exist")
+        else:
+            cmd = f"PowerShell (New-Object System.Net.WebClient).DownloadFile(" \
+                f"'{image_path}', 'C:\hyperv_img\{guest_name}.vhdx')"
+            ret, output = self.ssh.runcmd(cmd)
+            if not ret:
+                logger.info("succeeded to download hyperv image")
+            else:
+                raise FailException("Failed to download hyperv image")
+
+    def guest_exist(self, guest_name):
+        """
+        Check if the hyperv guest exists
+        :param guest_name: the name for the guest
+        :return: guest exists, return True, else, return False.
+        """
+        cmd = "PowerShell Get-VM"
+        ret, output = self.ssh.runcmd(cmd)
+        return not ret and guest_name in output
 
+    def guest_add(self, guest_name, image_path):
+        """
+        Create a new virtual machine.
+        :param guest_name: the name for the new virtual machine
+        :param image_path: the path for the guest image which from the remote web
+        :return: create successfully, return True, else, return False
+        """
+        if self.guest_exist(guest_name):
+            self.guest_del(guest_name)
+        self.guest_image(guest_name, image_path)
+        switch_name = self.host_switch()
+        if '8.' in guest_name:
+            options = f'-MemoryStartupBytes 2GB -SwitchName {switch_name} -Generation 2'
+        else:
+            options = f'-MemoryStartupBytes 1GB -SwitchName {switch_name} -Generation 1'
+        cmd = f"PowerShell New-VM -Name {guest_name} -VHDPath \"C:\hyperv_img\{guest_name}.vhdx\" {options}"
+        self.ssh.runcmd(cmd)
+        if self.guest_exist(guest_name):
+            logger.info("Succeeded to add hyperv guest")
+        else:
+            raise FailException("Failed to add hyperv guest")
+        if '8.' in guest_name:
+            cmd = f"PowerShell Set-VMFirmware -VMName {guest_name} -EnableSecureBoot off"
+            self.ssh.runcmd(cmd)
+        return self.guest_start(guest_name)
```

Comment:
I also think we don't need run the guest_start in the guest_add function, which can be done by users as requirements.

Reply:
Removed it.